### DISCUSSION TITLE
Render a single Dialog with conditional content from two different components

### DISCUSSION THREAD

**Discussion by riazjahangir** [ariakit/ariakit#1266] [Questions] [answered]
We're introducing a new popup Dialog on our site. The content of this popup will be different depending on what components are being rendered on the page -- e.g. if certain button components are rendered, we want the popup to include specific text.

If no special component is rendered, we want the popup to default to generic text for the user.

Our naive approach was to render the generic text Dialog in the common base layout shared by all our pages, and to render the specific text Dialog in the special component. However this doesn't seem to work -- here is a reproduction that illustrates the problem: https://codesandbox.io/s/gifted-framework-if72cc?file=/src/App.js

We do not see the popup showing the specific text when we are rendering the page with the specific component. 

Any thoughts on how to achieve this effect of a single "global" popup Dialog whose content depends on the components being rendered?

**Comment by diegohaz** [comment, marked as answer]
You can receive the `text` prop in your `BaseLayout` component.

```jsx
const BaseLayout = ({ className, children, text = "some generic content" }) => (
  <div className={className}>
    {children}
    <Popup text={text} />
  </div>
);

const GenericPage = () => (
  <BaseLayout className="generic">generic page</BaseLayout>
);

const ActionPage = () => (
  <BaseLayout className="action" text="action-specific content">
    action page
  </BaseLayout>
);
```

You can also receive the entire element:

```jsx
const BaseLayout = ({ className, children, popup = <Popup text="some generic content" /> }) => (
  <div className={className}>
    {children}
    {popup}
  </div>
);

const GenericPage = () => (
  <BaseLayout className="generic">generic page</BaseLayout>
);

const ActionPage = () => (
  <BaseLayout className="action" popup={<Popup text="action-specific content" />}>
    action page
  </BaseLayout>
);
```

**Reply by riazjahangir** [reply]
Hi @diegohaz ! That will work in my oversimplified example but really we have a deep component tree within `ActionPage`, and the content is owned by a component that is several descendants deep beneath `BaseLayout`.  To illustrate with a still-simplified example:
```
const ActionPage = () => (
  <BaseLayout>
    <Section1>
      <Section2/>
    </Section1>
  </BaseLayout>
)
```

```
const Section2 = () => {
  // This is the component that determines the popup content
}
```

Lifting up this content to pass as a prop to `BaseLayout` is indeed a clean solution for having a single global dialog, but here it's `Section2` that has the right context for what the popup content should be, so I'm struggling with how best to hand off this content between the components.

**Reply by diegohaz** [reply]
I'd use React Context then: https://codesandbox.io/s/boring-danilo-u7omwe?file=/src/App.js

I'm using Constate in that CodeSandbox, but the same works with pure React Context.

**Reply by riazjahangir** [reply]
Thanks @diegohaz ! I'd looked at React Context but it seems designed only to pass values _down_ a component tree, whereas we wanted to pass content back up to `BaseLayout`. Constate looks like the perfect solution here!
